components/apimgt/org.wso2.carbon.apimgt.common.analytics/src/main/java/org/wso2/carbon/apimgt/common/analytics/publishers/impl/AbstractRequestDataPublisher.java

```diff
@@ -43,27 +45,34 @@ public abstract class AbstractRequestDataPublisher implements RequestDataPublish
     @Override
     public void publish(Event analyticsEvent) {
 
-        CounterMetric counterMetric = this.getCounterMetric();
-        if (counterMetric == null) {
-            log.error("counterMetric cannot be null.");
+        Map<String, Object> dataMap = OBJECT_MAPPER.convertValue(analyticsEvent, MAP_TYPE_REFERENCE);
+        List<CounterMetric> multipleCounterMetrics = this.getMultipleCounterMetrics();
+        if (multipleCounterMetrics == null) {
+            log.error("All the counterMetrics are invalid. Event will be dropped.");
             return;
         }
 
-        Map<String, Object> dataMap = OBJECT_MAPPER.convertValue(analyticsEvent, MAP_TYPE_REFERENCE);
-        MetricEventBuilder builder = counterMetric.getEventBuilder();
-        for (Map.Entry<String, Object> entry : dataMap.entrySet()) {
-            try {
-                builder.addAttribute(entry.getKey(), entry.getValue());
-            } catch (MetricReportingException e) {
-                log.error("Error adding data to the event stream.", e);
-                return;
+        for (CounterMetric counterMetric : multipleCounterMetrics) {
+            if (counterMetric == null) {
+                log.error("counterMetric cannot be null.");
+            } else {
+                String counterMetricClassName = counterMetric.getClass().toString().
+                        replaceAll("[\r\n]", "").split(" ")[1];
+                log.info("Started adding data to counterMetric " + counterMetricClassName);
+                MetricEventBuilder builder = counterMetric.getEventBuilder();
+                for (Map.Entry<String, Object> entry : dataMap.entrySet()) {
+                    try {
+                        builder.addAttribute(entry.getKey(), entry.getValue());
+                    } catch (MetricReportingException e) {
+                        break;
+                    }
+                }
+                try {
+                    counterMetric.incrementCount(builder);
```

Comment:
This is still wrong. if we get a exception while adding an attribute we should just return with an error log saying publishing failed. if we could add everything we will only increment

Reply:
My bad. didnt see the outer loop. You will have to go back to earlier solution of having a variable. 
If exception is caught mark that var and break. then check the car and increment. sorry for the trouble

Reply:
Ack Tishan
No worries.